design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+        "version": "r207.kbase.3"
+    },
+    {
+        "product": "genome_stats",
+        "version': "r207.kbase.14"
+    }
+]
+```
+
+## Implementation
+
+### DB schema
+
+* The collections document will be essentially as represented above, with the keys / values
+  as described and a list of associated data products.
+* There will be two separate collections - `coll_all_versions` (or something) and
+  `coll_active_versions`.
+  * The former holds all versions of all collections, and the latter holds only the active
+    version of each collection.
+* Activating a version means:
+  * Pulling the collection version document from `coll_all_versions`
+  * Setting the activation date to the current date
+  * Overwriting the collection document in `coll_active_versions`.
+    * Atomic!
+* Alternatives:
+  * Tag the active version in `coll_active_version`.
+    * There's a fundamental problem with this approach in that it's non-atomic - at some point
+      either no versions will be tagged as active or > 1 will.
+    * This can be alleviated with transactions but in Arango [transactions are only reliable in
+      non-sharded environments](https://github.com/arangodb/arangodb/issues/11424).
+    * Furthermore, filtering data becomes much slower since all the inactive versions need to
+      be filtered out. Index size goes up because you need (at least) the activity state in
+      the equality part of the index. If you want to filter inactive versions, you need a
+      separate index without the activity state in the equality part of the index.
+    * Also generally more complex than the proposed schema.
+  * Have the service manage the collection version with a monotonically increasing version, like
+    `workspace_deluxe` or `sample_service`
+    * This approach is much more complex and is not necessary for the use case proposed here. It
+      also has the filtering drawback as mentioned above.
+
+### Endpoints
+
+* Only includes immediately relevant endpoints.
+
+#### List collections
+
+* For now returns all active collections based on the < 1000 collections assumption.
+
+```
+GET /collection
+```
+
+#### Save a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+POST /collection/{collection_id}/version/{version}
+
+BODY contains collection contents.
+```
+
+#### Activate a collection
+
+* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.
+
+```
+PUT /collection/{collection_id}/version/{version}/activate
+```
+
+#### Get a collection
+
+```
+GET /collection/{collection_id}
+```
+
+* Extendable to allow getting a specific version in the future.
+
+#### Data products
+
+* General endpoint structure, exact path and query params will depend on the data product
+
+```
+GET /collection/{collection_id}/data/{data_product_id}/...
+```
+* Extendable to allow getting a specific collection version (which specifies the data product
+  version) in the future.
```

Comment:
super minor nit, plurals seem to be more common in rest APIs to help differentiate between "functional" endpoints (like `.../activate`) and "object" endpoints (like `/collections`):

````suggestion
### Endpoints

* Only includes immediately relevant endpoints.

#### List collections

* For now returns all active collections based on the < 1000 collections assumption.

```
GET /collections
```

#### Save a collection

* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.

```
POST /collections/{collection_id}/versions/{version}

BODY contains collection contents.
```

#### Activate a collection

* Requires a token with the `COLLECTIONS_ADMIN` (or something) `auth2` role.

```
PUT /collections/{collection_id}/versions/{version}/activate
```

#### Get a collection

```
GET /collections/{collection_id}
```

* Extendable to allow getting a specific version in the future.

#### Data products

* General endpoint structure, exact path and query params will depend on the data product

```
GET /collections/{collection_id}/data/{data_product_id}/...
```
* Extendable to allow getting a specific collection version (which specifies the data product
  version) in the future.
````

Reply:
I don't really care one way or the other, but it looks like the plurals have it: https://stackoverflow.com/questions/6845772/should-i-use-singular-or-plural-name-convention-for-rest-resources

Reply:
anyway, done

Reply:
resolve as per @dauglyon